pagefaultgames/pokerogue
# [Enhancement] Add Move Header phase and attributes

## What are the changes?

This adds a new phase, `MoveHeaderPhase`, which takes place before all moves in a turn and applies all the `MoveHeaderAttr` attributes of a move.

## Why am I doing these changes?



At the moment, this acts as a partial fix to Focus Punch missing its on-cart messages (closes #898), but the overall intent is to create a framework so that Beak Blast and Shell Trap can be implemented (or reimplemented) in a more structured and intuitive way.

## What did change?


- `phases`: A new 'MoveHeaderPhase' is added and integrated into the game's turn sequence. With this new phase, the generic turn sequence (after all commands are selected) is as follows:
  1. If the player chooses to throw a ball, switch a Pokemon, or run away, those actions are resolved first.
  2. For each Pokemon that uses a move with a `MoveHeaderAttr`, a `MoveHeaderPhase` is added to the turn sequence. This phase applies all of the move's `MoveHeaderAttrs`, as long as the move itself can be used. If the user has a status that may block the use of the move, the status check is done here instead of in the `MovePhase`. `MoveHeaderPhases` are resolved in the same order as the turn's move order.
  3. `MovePhases` are added to the turn sequence for each move. If the user's status was previously checked in a `MoveHeaderPhase`, the status check here is ignored. Otherwise, this phase and all subsequent phases function in the same way as on upstream.
- `data/move`: `MoveHeaderAttr` is a new subclass of `MoveAttr` that encompasses all attributes that are applied in the `MoveHeaderPhase`. For now, the only `MoveHeaderAttr` implemented is `MessageHeaderAttr`, which queues a message, but more subclasses of `MoveHeaderAttr` can be created in the future to apply more effects, e.g. adding a battler tag to the user.
  - Focus Punch now uses a `MessageHeaderAttr` to show its initial message.
- `test/moves/focus_punch.test`: New unit test to ensure Focus Punch's effect timing is correct throughout the turn sequence

### Screenshots/Videos



Basic Example showing Focus Punch's new pre-move message:

https://github.com/pagefaultgames/pokerogue/assets/168692175/ca49eda0-341d-4269-9c5d-f7baf689fb26

## How to test the changes?




1. `npm run test focus_punch`
2. Use `src/overrides` to give the player Pokemon Focus Punch, then use `npm run start:dev` and observe the turn sequence in battle when using the move:
    1. What happens when using Focus Punch compared to other moves?
    2. What happens when one player Pokemon switches, and the other uses Focus Punch?
    3. What happens when one player Pokemon throws a ball or runs, and the other uses Focus Punch?

## Checklist
